File: python/periph/chips/power/ina3221.py
import struct


class INA3221Minimal:
    """INA3221 three-channel 26V current/voltage/power monitor — minimal interface.

    Reads bus voltage, shunt voltage, current, and power for each of the three
    channels with no configuration beyond the transport and shunt resistors.
    The chip's power-on default (all three channels on, continuous shunt+bus)
    is used without modification.

    Args:
        transport: Configured I2C or SMBus transport pointing at the device.
        r_shunt: Shunt resistor value in ohms. Pass a single float to apply
            the same value to all three channels, or a 3-element sequence
            (list/tuple) for per-channel values (default 0.1 ohms for all).
    """

    _REG_CONFIG   = 0x00
    _REG_SHUNT1   = 0x01
    _REG_BUS1     = 0x02
    _REG_SHUNT2   = 0x03
    _REG_BUS2     = 0x04
    _REG_SHUNT3   = 0x05
    _REG_BUS3     = 0x06
    _REG_MFR_ID   = 0xFE
    _REG_DIE_ID   = 0xFF

    _SHUNT_REGS = (_REG_SHUNT1, _REG_SHUNT2, _REG_SHUNT3)
    _BUS_REGS   = (_REG_BUS1, _REG_BUS2, _REG_BUS3)

    def __init__(self, transport, r_shunt=0.1):
        self._transport = transport
        if hasattr(r_shunt, '__iter__'):
            self._r_shunt = tuple(float(r_shunt[i]) for i in range(3))
        else:
            self._r_shunt = (float(r_shunt), float(r_shunt), float(r_shunt))

    def _write_reg(self, reg, value):
        self._transport.write(struct.pack('>BH', reg, value))

    def _read_reg(self, reg):
        return struct.unpack('>H', self._transport.write_read(bytes([reg]), 2))[0]

    def _read_reg_signed(self, reg):
        return struct.unpack('>h', self._transport.write_read(bytes([reg]), 2))[0]

    def _channel_valid(self, channel):
        if channel not in (1, 2, 3):
            raise ValueError('channel must be 1, 2, or 3')
        return channel

    def voltage(self, channel):
        """Read bus voltage for a channel.

        Args:
            channel: Channel number 1, 2, or 3.

        Returns:
            float: Bus voltage in volts.
        """
        ch = self._channel_valid(channel)
        raw = self._read_reg(self._BUS_REGS[ch - 1])
        return (raw >> 3) * 8e-3

    def shunt_voltage(self, channel):
        """Read differential shunt voltage for a channel.

        Args:
            channel: Channel number 1, 2, or 3.

        Returns:
            float: Shunt voltage in volts, signed.
        """
        ch = self._channel_valid(channel)
        raw = self._read_reg_signed(self._SHUNT_REGS[ch - 1])
        return raw * 5e-6

    def current(self, channel):
        """Read calculated current through the shunt for a channel.

        Args:
            channel: Channel number 1, 2, or 3.

        Returns:
            float: Current in amperes.
        """
        ch = self._channel_valid(channel)
        return self.shunt_voltage(ch) / self._r_shunt[ch - 1]

    def power(self, channel):
        """Read calculated power for a channel.

        Args:
            channel: Channel number 1, 2, or 3.

        Returns:
            float: Power in watts.
        """
        ch = self._channel_valid(channel)
        return self.voltage(ch) * self.current(ch)


class INA3221Full(INA3221Minimal):
    """INA3221 full interface — extends INA3221Minimal with configuration and alert support.

    Adds Configuration Register programming, channel enables, conversion-ready,
    per-channel critical and warning alerts, shunt-voltage summation, power-valid
    monitoring, reset, and shutdown/wake.

    Alert flag constants (from Mask/Enable register):
        CF1, CF2, CF3 — Channel 1/2/3 critical-alert flag
        WF1, WF2, WF3 — Channel 1/2/3 warning-alert flag
        SF            — Summation-alert flag
        PVF           — Power-valid flag
        TCF           — Timing-control flag
        CVRF          — Conversion-ready flag

    Mode constants:
        MODE_POWERDOWN      = 0
        MODE_SHUNT_TRIG     = 1
        MODE_BUS_TRIG       = 2
        MODE_SHUNT_BUS_TRIG = 3
        MODE_SHUNT_CONT     = 5
        MODE_BUS_CONT       = 6
        MODE_SHUNT_BUS_CONT = 7

    Args:
        transport: Configured I2C or SMBus transport pointing at the device.
        r_shunt: Shunt resistor value in ohms. Pass a single float to apply
            the same value to all three channels, or a 3-element sequence
            (list/tuple) for per-channel values (default 0.1 ohms for all).
    """

    CF1 = 0x0200
    CF2 = 0x0100
    CF3 = 0x0080
    SF  = 0x0040
    WF1 = 0x0020
    WF2 = 0x0010
    WF3 = 0x0008
    PVF = 0x0004
    TCF = 0x0002
    CVRF = 0x0001

    MODE_POWERDOWN      = 0
    MODE_SHUNT_TRIG     = 1
    MODE_BUS_TRIG       = 2
    MODE_SHUNT_BUS_TRIG = 3
    MODE_SHUNT_CONT     = 5
    MODE_BUS_CONT       = 6
    MODE_SHUNT_BUS_CONT = 7

    _REG_CH1_CRIT   = 0x07
    _REG_CH1_WARN   = 0x08
    _REG_CH2_CRIT   = 0x09
    _REG_CH2_WARN   = 0x0A
    _REG_CH3_CRIT   = 0x0B
    _REG_CH3_WARN   = 0x0C
    _REG_SUM        = 0x0D
    _REG_SUM_LIMIT  = 0x0E
    _REG_MASK_EN    = 0x0F
    _REG_PV_UPPER   = 0x10
    _REG_PV_LOWER   = 0x11

    _CRIT_REGS  = (_REG_CH1_CRIT, _REG_CH2_CRIT, _REG_CH3_CRIT)
    _WARN_REGS  = (_REG_CH1_WARN, _REG_CH2_WARN, _REG_CH3_WARN)

    def __init__(self, transport, r_shunt=0.1):
        super().__init__(transport, r_shunt)
        self._mode = 0x07

    def configure(self, avg=0, vbus_ct=4, vsh_ct=4, mode=7):
        """Write the Configuration Register.

        Preserves channel-enable bits (CH1en, CH2en, CH3en).

        Args:
            avg: Averaging count selector 0-7 (0=1 sample, 7=1024 samples).
            vbus_ct: Bus voltage conversion time selector 0-7 (default 4=1.1 ms).
            vsh_ct: Shunt voltage conversion time selector 0-7 (default 4=1.1 ms).
            mode: Operating mode (default 7=shunt+bus continuous).
        """
        cfg = self._read_reg(self._REG_CONFIG)
        config = ((avg & 0x07) << 9) | ((vbus_ct & 0x07) << 6) | ((vsh_ct & 0x07) << 3) | (mode & 0x07)
        config |= cfg & 0x7000
        self._mode = mode & 0x07
        self._write_reg(self._REG_CONFIG, config)

    def enable_channel(self, channel, enabled):
        """Enable or disable a channel.

        Args:
            channel: Channel number 1, 2, or 3.
            enabled: True to enable, False to disable.
        """
        ch = self._channel_valid(channel)
        cfg = self._read_reg(self._REG_CONFIG)
        bit = 14 - (ch - 1)
        if enabled:
            cfg |= (1 << bit)
        else:
            cfg &= ~(1 << bit)
        self._write_reg(self._REG_CONFIG, cfg)

    def channel_enabled(self, channel):
        """Read whether a channel is enabled.

        Args:
            channel: Channel number 1, 2, or 3.

        Returns:
            bool: True if the channel is enabled.
        """
        ch = self._channel_valid(channel)
        cfg = self._read_reg(self._REG_CONFIG)
        bit = 14 - (ch - 1)
        return bool(cfg & (1 << bit))

    def conversion_ready(self):
        """Read the Conversion Ready Flag (CVRF).

        Returns:
            bool: True if a conversion completed.
        """
        return bool(self._read_reg(self._REG_MASK_EN) & self.CVRF)

    def set_critical_alert(self, channel, limit_v, latch=False):
        """Set the critical-alert limit for a channel.

        Args:
            channel: Channel number 1, 2, or 3.
            limit_v: Voltage limit in volts.
            latch: If True, use latched mode (default False).
        """
        ch = self._channel_valid(channel)
        raw = (int(limit_v / 40e-6) << 3) & 0xFFF8
        self._write_reg(self._CRIT_REGS[ch - 1], raw)
        cfg = self._read_reg(self._REG_MASK_EN)
        if latch:
            cfg |= 0x0400
        else:
            cfg &= ~0x0400
        self._write_reg(self._REG_MASK_EN, cfg)

    def set_warning_alert(self, channel, limit_v, latch=False):
        """Set the warning-alert limit for a channel.

        Args:
            channel: Channel number 1, 2, or 3.
            limit_v: Voltage limit in volts.
            latch: If True, use latched mode (default False).
        """
        ch = self._channel_valid(channel)
        raw = (int(limit_v / 40e-6) << 3) & 0xFFF8
        self._write_reg(self._WARN_REGS[ch - 1], raw)
        cfg = self._read_reg(self._REG_MASK_EN)
        if latch:
            cfg |= 0x0800
        else:
            cfg &= ~0x0800
        self._write_reg(self._REG_MASK_EN, cfg)

    def alert_flags(self):
        """Read the Mask/Enable Register.

        Reading this register clears the latched alert flags (CF1/CF2/CF3,
        WF1/WF2/WF3, SF) when latch mode is enabled.

        Returns:
            int: Raw 16-bit Mask/Enable register value.
        """
        return self._read_reg(self._REG_MASK_EN)

    def set_summation_channels(self, channels, limit_v):
        """Configure the shunt-voltage summation function.

        Args:
            channels: List of channel numbers to sum (e.g. [1, 2, 3]).
            limit_v: Shunt-voltage sum limit in volts.
        """
        cfg = self._read_reg(self._REG_MASK_EN)
        cfg &= ~0xE000
        for ch in channels:
            self._channel_valid(ch)
            cfg |= 1 << (15 - (ch - 1))
        self._write_reg(self._REG_MASK_EN, cfg)
        raw = (int(limit_v / 40e-6) << 1) & 0xFFFE
        self._write_reg(self._REG_SUM_LIMIT, raw)

    def summation_value(self):
        """Read the shunt-voltage sum.

        Returns:
            float: Sum of selected channels' shunt voltages in volts.
        """
        raw = self._read_reg_signed(self._REG_SUM)
        return raw * 5e-6

    def set_power_valid_limits(self, upper_v, lower_v):
        """Set the Power-Valid upper and lower voltage limits.

        Args:
            upper_v: Upper bus voltage limit in volts.
            lower_v: Lower bus voltage limit in volts.
        """
        raw_upper = (int(upper_v / 8e-3) << 3) & 0xFFF8
        raw_lower = (int(lower_v / 8e-3) << 3) & 0xFFF8
        self._write_reg(self._REG_PV_UPPER, raw_upper)
        self._write_reg(self._REG_PV_LOWER, raw_lower)

    def power_valid(self):
        """Read the Power-Valid flag (PVF).

        Returns:
            bool: True if all enabled bus voltages are within the PV limits.
        """
        return bool(self._read_reg(self._REG_MASK_EN) & self.PVF)

    def shutdown(self):
        """Enter power-down mode and save the current mode for wake()."""
        cfg = self._read_reg(self._REG_CONFIG)
        self._mode = cfg & 0x07
        self._write_reg(self._REG_CONFIG, cfg & 0xFFF8)

    def wake(self):
        """Restore the operating mode saved by shutdown()."""
        cfg = self._read_reg(self._REG_CONFIG)
        self._write_reg(self._REG_CONFIG, (cfg & 0xFFF8) | self._mode)

    def reset(self):
        """Reset all registers to power-on defaults."""
        self._write_reg(self._REG_CONFIG, 0x8000)

    def manufacturer_id(self):
        """Read the Manufacturer ID register.

        Returns:
            int: Manufacturer ID; expect 0x5449 (Texas Instruments).
        """
        return self._read_reg(self._REG_MFR_ID)

    def die_id(self):
        """Read the Die ID register.

        Returns:
            int: Die revision ID; expect 0x3220.
        """
        return self._read_reg(self._REG_DIE_ID)

File: go/periph/chips/power/ina3221.go
// Package power contains drivers for Power management chips.
package power

import (
	"fmt"

	"github.com/tuhde/Periph/go/periph/transport"
)

// INA3221 register addresses.
const (
	ina3221RegConfig   uint8 = 0x00
	ina3221RegShunt1   uint8 = 0x01
	ina3221RegBus1     uint8 = 0x02
	ina3221RegShunt2   uint8 = 0x03
	ina3221RegBus2     uint8 = 0x04
	ina3221RegShunt3   uint8 = 0x05
	ina3221RegBus3     uint8 = 0x06
	ina3221RegCh1Crit  uint8 = 0x07
	ina3221RegCh1Warn  uint8 = 0x08
	ina3221RegCh2Crit  uint8 = 0x09
	ina3221RegCh2Warn  uint8 = 0x0A
	ina3221RegCh3Crit  uint8 = 0x0B
	ina3221RegCh3Warn  uint8 = 0x0C
	ina3221RegSum      uint8 = 0x0D
	ina3221RegSumLimit uint8 = 0x0E
	ina3221RegMaskEn   uint8 = 0x0F
	ina3221RegPVUpper  uint8 = 0x10
	ina3221RegPVLower  uint8 = 0x11
	ina3221RegMfrID    uint8 = 0xFE
	ina3221RegDieID    uint8 = 0xFF
)

// INA3221 Mask/Enable alert flag bit positions, matching the chip's
// register layout. Decoded from INA3221Full.AlertFlags.
const (
	// CF1 is the channel-1 critical-alert flag bit.
	CF1 uint16 = 0x0200
	// CF2 is the channel-2 critical-alert flag bit.
	CF2 uint16 = 0x0100
	// CF3 is the channel-3 critical-alert flag bit.
	CF3 uint16 = 0x0080
	// SF is the summation-alert flag bit.
	SF uint16 = 0x0040
	// WF1 is the channel-1 warning-alert flag bit.
	WF1 uint16 = 0x0020
	// WF2 is the channel-2 warning-alert flag bit.
	WF2 uint16 = 0x0010
	// WF3 is the channel-3 warning-alert flag bit.
	WF3 uint16 = 0x0008
	// PVF is the power-valid flag bit.
	PVF uint16 = 0x0004
	// TCF is the timing-control flag bit.
	TCF uint16 = 0x0002
	// CVRF is the conversion-ready flag bit.
	CVRF uint16 = 0x0001
)

// INA3221 operating-mode constants share their lower 3 bits with the
// INA219 and are declared in ina219.go (MODE_POWERDOWN through
// MODE_SHUNT_BUS_CONT). The INA3221 spec treats MODE 100 as a second
// power-down encoding — callers can use either MODE_POWERDOWN (= 0) or
// MODE_ADC_OFF (= 4) to power the chip down.

// ina3221ShuntRegs maps channel number (1–3) to its shunt voltage register.
var ina3221ShuntRegs = [3]uint8{ina3221RegShunt1, ina3221RegShunt2, ina3221RegShunt3}

// ina3221BusRegs maps channel number (1–3) to its bus voltage register.
var ina3221BusRegs = [3]uint8{ina3221RegBus1, ina3221RegBus2, ina3221RegBus3}

// ina3221CritRegs maps channel number (1–3) to its critical-alert limit register.
var ina3221CritRegs = [3]uint8{ina3221RegCh1Crit, ina3221RegCh2Crit, ina3221RegCh3Crit}

// ina3221WarnRegs maps channel number (1–3) to its warning-alert limit register.
var ina3221WarnRegs = [3]uint8{ina3221RegCh1Warn, ina3221RegCh2Warn, ina3221RegCh3Warn}

// ina3221ValidChannel converts a 1-based channel number into a 0-based
// index, returning an error for any value outside 1–3.
func ina3221ValidChannel(channel uint8) (uint8, error) {
	if channel < 1 || channel > 3 {
		return 0, fmt.Errorf("ina3221: channel must be 1, 2, or 3, got %d", channel)
	}
	return channel - 1, nil
}

// INA3221Minimal is the INA3221 three-channel 26V current/voltage/power
// monitor driver — minimal interface.
//
// Communicates over I²C at up to 400 kHz fast mode. The INA3221 has no
// Calibration, Current, or Power registers — current and power are
// computed in software from the shunt voltage and the user-supplied
// shunt resistance.
//
// The chip's power-on defaults are used (all three channels enabled,
// continuous shunt+bus, max range, no averaging) — the Minimal driver
// writes nothing at construction.
type INA3221Minimal struct {
	transport transport.Transport
	rShunt    [3]float32
}

// NewINA3221Minimal creates a new INA3221Minimal with a single shunt
// resistance value applied to all three channels.
//
// transport must be a configured I²C transport bound to the device's
// 7-bit address (0x40–0x43, default 0x40).
//
// rShunt is the shunt resistor value in ohms, applied to all three channels.
func NewINA3221Minimal(transport transport.Transport, rShunt float32) (*INA3221Minimal, error) {
	return &INA3221Minimal{
		transport: transport,
		rShunt:    [3]float32{rShunt, rShunt, rShunt},
	}, nil
}

// NewINA3221MinimalPerChannel creates a new INA3221Minimal with
// per-channel shunt resistance values.
//
// rShunt is a 3-element array of shunt resistor values in ohms, one per
// channel (index 0 = channel 1, index 1 = channel 2, index 2 = channel 3).
func NewINA3221MinimalPerChannel(transport transport.Transport, rShunt [3]float32) (*INA3221Minimal, error) {
	return &INA3221Minimal{transport: transport, rShunt: rShunt}, nil
}

// writeReg writes a 16-bit value to a register (big-endian, pointer byte
// followed by two data bytes).
func (d *INA3221Minimal) writeReg(reg uint8, value uint16) error {
	return d.transport.Write([]byte{reg, byte(value >> 8), byte(value & 0xFF)})
}

// readReg reads a 16-bit unsigned register value.
func (d *INA3221Minimal) readReg(reg uint8) (uint16, error) {
	buf, err := d.transport.WriteRead([]byte{reg}, 2)
	if err != nil {
		return 0, err
	}
	return uint16(buf[0])<<8 | uint16(buf[1]), nil
}

// readRegSigned reads a 16-bit register value and returns it as signed.
func (d *INA3221Minimal) readRegSigned(reg uint8) (int16, error) {
	v, err := d.readReg(reg)
	if err != nil {
		return 0, err
	}
	return int16(v), nil
}

// Voltage reads the bus voltage for a channel. LSB = 8 mV; the raw
// register is left-aligned by 3, so right-shifting by 3 then scaling
// gives the same result as multiplying the signed 16-bit raw value by 1 mV.
func (d *INA3221Minimal) Voltage(channel uint8) (float32, error) {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return 0, err
	}
	raw, err := d.readReg(ina3221BusRegs[idx])
	if err != nil {
		return 0, err
	}
	return float32(raw>>3) * 8e-3, nil
}

// ShuntVoltage reads the differential shunt voltage for a channel.
// LSB = 40 µV. The raw register stores a 13-bit signed left-aligned
// value; multiplying the signed 16-bit raw value by 5 µV is mathematically
// equivalent to an explicit sign-extending right-shift by 3.
func (d *INA3221Minimal) ShuntVoltage(channel uint8) (float32, error) {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return 0, err
	}
	raw, err := d.readRegSigned(ina3221ShuntRegs[idx])
	if err != nil {
		return 0, err
	}
	return float32(raw) * 5e-6, nil
}

// Current reads the calculated current for a channel, computed as
// shunt_voltage / r_shunt[channel].
func (d *INA3221Minimal) Current(channel uint8) (float32, error) {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return 0, err
	}
	sv, err := d.ShuntVoltage(channel)
	if err != nil {
		return 0, err
	}
	return sv / d.rShunt[idx], nil
}

// Power reads the calculated power for a channel, computed as
// voltage * current.
func (d *INA3221Minimal) Power(channel uint8) (float32, error) {
	v, err := d.Voltage(channel)
	if err != nil {
		return 0, err
	}
	c, err := d.Current(channel)
	if err != nil {
		return 0, err
	}
	return v * c, nil
}

// INA3221Full is the INA3221 driver — full interface. Extends
// INA3221Minimal with Configuration Register access, channel enables,
// per-channel critical and warning alerts, shunt-voltage summation,
// power-valid monitoring, reset, and shutdown/wake.
type INA3221Full struct {
	*INA3221Minimal
	mode uint8
}

// NewINA3221Full creates a new INA3221Full with a single shunt resistance
// value applied to all three channels. Same arguments as
// NewINA3221Minimal.
func NewINA3221Full(transport transport.Transport, rShunt float32) (*INA3221Full, error) {
	m, err := NewINA3221Minimal(transport, rShunt)
	if err != nil {
		return nil, err
	}
	return &INA3221Full{INA3221Minimal: m, mode: 0x07}, nil
}

// NewINA3221FullPerChannel creates a new INA3221Full with per-channel
// shunt resistance values.
func NewINA3221FullPerChannel(transport transport.Transport, rShunt [3]float32) (*INA3221Full, error) {
	m, err := NewINA3221MinimalPerChannel(transport, rShunt)
	if err != nil {
		return nil, err
	}
	return &INA3221Full{INA3221Minimal: m, mode: 0x07}, nil
}

// Configure writes the Configuration Register, preserving the channel-
// enable bits (CH1en, CH2en, CH3en) in bits 14:12.
//
// avg is the averaging count selector 0–7 (0 = 1 sample, 7 = 1024).
// vbusCT is the bus voltage conversion time selector 0–7 (default 4 = 1.1 ms).
// vshCT is the shunt voltage conversion time selector 0–7 (default 4 = 1.1 ms).
// mode is the operating mode 0–7 (default 7 = shunt+bus continuous).
func (d *INA3221Full) Configure(avg, vbusCT, vshCT, mode uint8) error {
	cfg, err := d.readReg(ina3221RegConfig)
	if err != nil {
		return err
	}
	config := (uint16(avg&0x07) << 9) |
		(uint16(vbusCT&0x07) << 6) |
		(uint16(vshCT&0x07) << 3) |
		uint16(mode&0x07)
	config |= cfg & 0x7000
	d.mode = mode & 0x07
	return d.writeReg(ina3221RegConfig, config)
}

// EnableChannel sets or clears the CH<n>en bit in the Configuration
// Register. Disabled channels are bypassed in the measurement sequence.
func (d *INA3221Full) EnableChannel(channel uint8, enabled bool) error {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return err
	}
	cfg, err := d.readReg(ina3221RegConfig)
	if err != nil {
		return err
	}
	bit := uint16(1) << uint(14-idx)
	if enabled {
		cfg |= bit
	} else {
		cfg &^= bit
	}
	return d.writeReg(ina3221RegConfig, cfg)
}

// ChannelEnabled reports whether a channel is enabled.
func (d *INA3221Full) ChannelEnabled(channel uint8) (bool, error) {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return false, err
	}
	cfg, err := d.readReg(ina3221RegConfig)
	if err != nil {
		return false, err
	}
	bit := uint16(1) << uint(14-idx)
	return cfg&bit != 0, nil
}

// ConversionReady reads the CVRF (Conversion Ready Flag) bit from the
// Mask/Enable Register. CVRF clears on a write to the Configuration
// Register (except for power-down) or on a read of Mask/Enable.
func (d *INA3221Full) ConversionReady() (bool, error) {
	v, err := d.readReg(ina3221RegMaskEn)
	if err != nil {
		return false, err
	}
	return v&CVRF != 0, nil
}

// SetCriticalAlert writes the critical-alert limit for a channel and
// updates the CEN (Critical-alert latch) bit in the Mask/Enable Register.
//
// limitV is the shunt-voltage threshold in volts. latch enables latched
// alert mode when true.
func (d *INA3221Full) SetCriticalAlert(channel uint8, limitV float32, latch bool) error {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return err
	}
	raw := (uint16(limitV/40e-6) << 3) & 0xFFF8
	if err := d.writeReg(ina3221CritRegs[idx], raw); err != nil {
		return err
	}
	cfg, err := d.readReg(ina3221RegMaskEn)
	if err != nil {
		return err
	}
	if latch {
		cfg |= 0x0400
	} else {
		cfg &^= 0x0400
	}
	return d.writeReg(ina3221RegMaskEn, cfg)
}

// SetWarningAlert writes the warning-alert limit for a channel and
// updates the WEN (Warning-alert latch) bit in the Mask/Enable Register.
//
// limitV is the shunt-voltage threshold in volts. latch enables latched
// alert mode when true.
func (d *INA3221Full) SetWarningAlert(channel uint8, limitV float32, latch bool) error {
	idx, err := ina3221ValidChannel(channel)
	if err != nil {
		return err
	}
	raw := (uint16(limitV/40e-6) << 3) & 0xFFF8
	if err := d.writeReg(ina3221WarnRegs[idx], raw); err != nil {
		return err
	}
	cfg, err := d.readReg(ina3221RegMaskEn)
	if err != nil {
		return err
	}
	if latch {
		cfg |= 0x0800
	} else {
		cfg &^= 0x0800
	}
	return d.writeReg(ina3221RegMaskEn, cfg)
}

// AlertFlags reads the raw 16-bit Mask/Enable Register. Reading the
// register clears the latched CF/WF/SF flags when latch mode is enabled.
func (d *INA3221Full) AlertFlags() (uint16, error) {
	return d.readReg(ina3221RegMaskEn)
}

// SetSummationChannels configures the shunt-voltage summation function.
// The selected channels' single-conversion shunt readings are summed
// into the Shunt-Voltage Sum register and compared against limitV.
//
// The summation function only makes sense when all selected channels
// use the same shunt resistance — the limit is in volts, not amperes.
func (d *INA3221Full) SetSummationChannels(channels []uint8, limitV float32) error {
	cfg, err := d.readReg(ina3221RegMaskEn)
	if err != nil {
		return err
	}
	cfg &^= 0xE000
	for _, ch := range channels {
		idx, err := ina3221ValidChannel(ch)
		if err != nil {
			return err
		}
		cfg |= uint16(1) << uint(15-idx)
	}
	if err := d.writeReg(ina3221RegMaskEn, cfg); err != nil {
		return err
	}
	raw := (uint16(limitV/40e-6) << 1) & 0xFFFE
	return d.writeReg(ina3221RegSumLimit, raw)
}

// SummationValue reads the Shunt-Voltage Sum Register and returns the
// sum of the selected channels' shunt voltages in volts.
func (d *INA3221Full) SummationValue() (float32, error) {
	raw, err := d.readRegSigned(ina3221RegSum)
	if err != nil {
		return 0, err
	}
	return float32(raw) * 5e-6, nil
}

// SetPowerValidLimits writes the Power-Valid upper and lower bus-voltage
// thresholds. The PV pin asserts high when all enabled bus-voltage
// readings exceed upperV, and de-asserts when any drops below lowerV.
func (d *INA3221Full) SetPowerValidLimits(upperV, lowerV float32) error {
	rawUpper := (uint16(upperV/8e-3) << 3) & 0xFFF8
	if err := d.writeReg(ina3221RegPVUpper, rawUpper); err != nil {
		return err
	}
	rawLower := (uint16(lowerV/8e-3) << 3) & 0xFFF8
	return d.writeReg(ina3221RegPVLower, rawLower)
}

// PowerValid reads the PVF (Power-Valid Flag) bit from the Mask/Enable
// Register. True when all enabled bus voltages are within the PV limits.
func (d *INA3221Full) PowerValid() (bool, error) {
	v, err := d.readReg(ina3221RegMaskEn)
	if err != nil {
		return false, err
	}
	return v&PVF != 0, nil
}

// Shutdown enters power-down mode (MODE = 000), saving the current mode
// for Wake.
func (d *INA3221Full) Shutdown() error {
	cfg, err := d.readReg(ina3221RegConfig)
	if err != nil {
		return err
	}
	d.mode = uint8(cfg & 0x07)
	return d.writeReg(ina3221RegConfig, cfg&0xFFF8)
}

// Wake restores the operating mode saved by Shutdown.
func (d *INA3221Full) Wake() error {
	cfg, err := d.readReg(ina3221RegConfig)
	if err != nil {
		return err
	}
	return d.writeReg(ina3221RegConfig, (cfg&0xFFF8)|uint16(d.mode))
}

// Reset sets the RST bit, restoring all registers to power-on defaults
// (including alert limits and the Mask/Enable register). The caller must
// re-program any user-set limits after this call.
func (d *INA3221Full) Reset() error {
	return d.writeReg(ina3221RegConfig, 0x8000)
}

// ManufacturerID reads the Manufacturer ID Register. Expect 0x5449
// (Texas Instruments, ASCII "TI"). Same value as the INA226.
func (d *INA3221Full) ManufacturerID() (uint16, error) {
	return d.readReg(ina3221RegMfrID)
}

// DieID reads the Die ID Register. Expect 0x3220.
func (d *INA3221Full) DieID() (uint16, error) {
	return d.readReg(ina3221RegDieID)
}
